CI Lint check for Enso code

radeusgd opened the pull request:
### Pull Request Description

Starting by enabling the static type analysis as a lint step. In the future we may also add things like formatting using https://github.com/enso-org/ensolint

Closes #12775

To run the check locally, you can use `sbt lintEnso`.

To lint just a single project, you can use `lintEnso <path>` or `lintEnso <name>`. Matching by name looks for matching projects in `test` and `distribution/lib/Standard` directories, e.g. you can run `lintEnso Base` to lint the `Standard.Base` library or e.g. `lintEnso Table_Tests`.

### Important Notes

<!--
- Mention important elements of the design.
- Mention any notable changes to APIs.
-->

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [ ] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [x] Unit tests have been written where possible.
- [ ] If meaningful changes were made to logic or tests affecting Enso Cloud integration in the libraries, 
      or the Snowflake database integration, a run of the [Extra Tests](https://github.com/enso-org/enso/actions/workflows/extra-nightly-tests.yml) has been scheduled.
  - If applicable, it is suggested to paste a link to a successful run of the Extra Tests.

radeusgd committed:
implement -Werror

radeusgd committed:
fix

radeusgd committed:
adding Lint check

radeusgd committed:
fix message

radeusgd committed:
adding cmd - stub

radeusgd committed:
adding running lints in Rust

radeusgd committed:
Revert "adding running lints in Rust"

This reverts commit 837785a9f49168a451d493aa74c5b89903e3ffc8.

radeusgd committed:
add lintEnso command

radeusgd committed:
fix working directory for run to avoid a warning

radeusgd committed:
scalafmt

radeusgd committed:
add `run backend lint` command

radeusgd committed:
fixes - log, flags

radeusgd committed:
flag did not help

radeusgd committed:
regenerate workflows

kazcw commented:
There is already a `./run libraries lint` that currently just runs `check_syntax` on everything that looks like an Enso file--it should probably be combined with the new command. Or eliminated, if the new linter can report syntax errors and parser warnings.

radeusgd commented:
> There is already a `./run libraries lint` that currently just runs `check_syntax` on everything that looks like an Enso file--it should probably be combined with the new command. Or eliminated, if the new linter can report syntax errors and parser warnings.

Thanks, I was not aware. Yes that seems redundant, the compiler should also report syntax errors. I will probably remove the `libraries lint` one, or rename `backend lint` to it.

radeusgd committed:
refactor 1

radeusgd committed:
refactor 2

radeusgd committed:
GH annotations

radeusgd committed:
comment

radeusgd committed:
relativize paths

radeusgd committed:
log more informative error

radeusgd committed:
fix - env var may be missing

radeusgd committed:
test warnings

radeusgd committed:
Revert "test warnings"

This reverts commit e37c463da1bd702844201d815ba7b6671bcd9a35.

radeusgd committed:
include warning annotations

radeusgd committed:
delete Exploratory_Benchmarks project

It was not maintained and was relying on older structure of benchmarks. We can bring it back from history if we need it.

radeusgd committed:
make the `Internal` detection a bit more flexible

radeusgd committed:
fix weird behaviour: Builder is on Vector module, but somehow is available as Vector.Builder on Vector _type_. To avoid this we just import it explicitly.

radeusgd committed:
refactor, use aggregate projects to avoid recompiling

radeusgd committed:
fix

radeusgd committed:
fix namespaces

radeusgd committed:
add temp dirs to gitignore

radeusgd commented:
The Lint is already proving its worth - it has found out that one of the functions was not fully applied, thus some tests were no longer executed - probably since a change in October 2024.

![image](https://github.com/user-attachments/assets/72bf2523-c604-4dc1-adef-5363286ad5e2)

radeusgd committed:
fix missing argument - now the Name_Length_Limits_Spec is running

radeusgd committed:
singleton type is more static so let's use static syntax for these methods

radeusgd committed:
move `backend lint` to `libraries lint`

radeusgd committed:
Merge branch 'develop' into wip/rw/type-analysis-on-ci

radeusgd committed:
fmt

radeusgd commented on test/Exploratory_Benchmarks/README.md:
The `Exploratory_Benchmarks` project was no longer maintained, it triggered many warnings about bad types when calling `Bench.measure` probably because it was not updated after the benchmark tool was changed.

I don't think keeping it in repo is still worth it. The experiments were done, if we want to redo them now they'd need to be modified significantly. Ongoing maintenance of them is not worth it - they were experiments and not full benchmarks.

We can always bring them back from history if we need.

Akirathan replied:
I agree. Let's remove `Exploratory_Benchmarks`.

JaroslavTulach commented on .gitignore:
```diff
@@ -166,6 +166,7 @@ distribution/lib/*/*/*/manifest.yaml
 
 test/Google_Api_Test/data/secret.json
```
The description mentions

> In the future we may also add things like formatting using https://github.com/enso-org/ensolint

- I don't think we want to reference such a repository in this PR description.
- Personally I believe there is no point in adding yet another technology stack - JVM should be enough
- We already have more than enough technologies/languges in the system

JaroslavTulach replied:
> it provides devs an easy to run command - sbt lintEnso to run a check

This is the info that should be in the PR description!

JaroslavTulach commented on build.sbt:
```diff
@@ -1,18 +1,20 @@
 import sbt.Keys.{libraryDependencies, scalacOptions}
 import sbt.addCompilerPlugin
-import sbt.complete.DefaultParsers._
+import sbt.complete.DefaultParsers.*
```
Can we avoid non-semantical changes based only one one's IDE setup?

radeusgd replied:
Sorry, I did not notice, I will try to revert it.

radeusgd replied:
I used `scalafmtSbt` and thought it will deal with it.

radeusgd replied:
I reverted the change in 79f94dcba7c46e21d472da2ab607f09ace448380

hubertp commented on project/EnsoLint.scala:
```diff
@@ -0,0 +1,123 @@
+          if (foundByName.length == 1)
+            foundByName.head
+          else
+            EnsoProjects.ofPath(Path.of(name))
```
```suggestion
        val project =
          foundByName match {
            case proj :: Nil => proj
            case _             => EnsoProjects.ofPath(Path.of(name))
          }
```

hubertp commented on project/EnsoProjects.scala:
```diff
@@ -0,0 +1,64 @@
+object EnsoProjects {
+  case class Project(namespace: Option[String], name: String, path: Path) {
+    def usesPrivateAccess: Boolean =
+      path.getParent.getFileName.toString == "test" && name.contains(
```
`.toString` obsolete?

hubertp replied:
Feels like a regex may be more appropriate here

radeusgd replied:
> `.toString` obsolete?

![image](https://github.com/user-attachments/assets/559b57d7-ff5d-41ea-ac16-81138aefbcb6)

> Feels like a regex may be more appropriate here
I've used the same logic that we had elsewhere. I will try to make it a common method. IMO regex is an overkill for such a simple check - the two checks are more readable than something like `_Internal_(.*_)+Tests`

JaroslavTulach commented on engine/runner/src/main/java/org/enso/runner/Main.java:
```diff
@@ -93,6 +93,7 @@ public class Main {
   private static final String NO_READ_IR_CACHES_OPTION = "no-read-ir-caches";
   private static final String DISABLE_PRIVATE_CHECK_OPTION = "disable-private-check";
   private static final String ENABLE_STATIC_ANALYSIS_OPTION = "enable-static-analysis";
+  private static final String TREAT_WARNINGS_AS_ERRORS_OPTION = "treat-warnings-as-errors";
```
- Adding this option and fixing #11844 is great.
- I don't like the CLI option name however
- it is too verbose for my taste
- what do other tools do?
    - `gcc` uses `-Werror`
    - `scalac` uses `-Werror`
    - `javac` uses `-Werror`
    - `ghc` uses `-Werror`
- wouldn't it be better to "follow the crowd"?

radeusgd replied:
Ok, I'm happy to rename it.

JaroslavTulach commented on engine/common/src/main/java/org/enso/common/ContextFactory.java:
```diff
@@ -120,6 +121,11 @@ public ContextFactory enableStaticAnalysis(boolean enableStaticAnalysis) {
     return this;
   }
 
+  public ContextFactory treatWarningsAsErrors(boolean treatWarningsAsErrors) {
```
- Using (a little bit verbose) `treatWarningsAsErrors` internally is fine
- It is good when the name is self describing
- IDEs help us type it with code completion
- just the CLI option is what I am concerned with as expressed in another comment

hubertp commented on test/Table_Tests/src/Database/Common/Common_Spec.enso:
```diff
@@ -154,7 +154,7 @@ add_specs (suite_builder : Suite_Builder) (prefix : Text) (create_connection_fn
 
     Default_Ordering_Spec.add_specs suite_builder prefix create_connection_fn setup
-    Names_Length_Limits_Spec.add_specs suite_builder prefix create_connection_fn
+    Names_Length_Limits_Spec.add_specs suite_builder prefix create_connection_fn setup
```
nice

JaroslavTulach commented on engine/runtime-compiler/src/main/scala/org/enso/compiler/Compiler.scala:
```diff
@@ -1119,7 +1119,7 @@ class Compiler(
           context.formatDiagnostic(compilerModule, diag, isOutputRedirected)
         printDiagnostic(formattedDiag.getMessage)
-        if (diag.isInstanceOf[Error]) {
+        if (diag.isInstanceOf[Error] || config.treatWarningsAsErrors) {
```
Radek, have you managed to create some test for this new flag?

radeusgd replied:
Added in 6e013df49224cd0278ff188c03bd36f36118b01c

JaroslavTulach commented on project/EnsoLint.scala:
```diff
@@ -0,0 +1,123 @@
+    }
+  }
+
+  private def runAll(projects: Seq[EnsoProjects.Project]): Boolean = {
```
- Quite a lot of logic in `sbt`
- It'd be better to make it a simpler wrapper around non-Scala, non-sbt code
- hopefully with #12937 all this extra code could be deleted
- On the other hand, this _"temporary project"_ is a clever workaround to the [possible problems](https://github.com/enso-org/enso/issues/12937#issuecomment-2826950933) ahead

Akirathan commented on project/DistributionPackage.scala:
```diff
@@ -419,14 +425,23 @@ object DistributionPackage {
+    *
+    * It will be the argument following `--in-project`, `--run` or `--compile`.
+    */
+  private def findProjectPath(args: Seq[String]): Option[String] = {
```
Note that once #12906 is integrated (should be soon), the `--in-project` cmdline option will be kind of obsolete.

radeusgd replied:
Do you mean you are removing `--in-project` option there?

Then, once merged we can remove it from here too.

Akirathan reviewed (changes requested):
I don't like duplicating logic for finding projects inside sbt, and I don't like the magic with aggregating the projects into a single one. I think it would be better to introduce a new sbt project `runtime-lints` that would contain unit tests that would locate all the libraries and run static type checks from them. This `runtime-lints` project could be excluded from JVM tests (so that they are not run from `sbt test` command). The CI command would be then something like `sbt runtime-lints/test`.

Moreover, the `includeGithubAnnotation` functionality could then be moved from `DiagnosticFormatter` (where it does not belong), into `runtime-lints/test` sources.

TL;DR; The logic for finding projects, and executing them, is already present all over the place in `runtime-integration-tests` for inspiration. Let's not introduce anymore sbt magic, please.

**EDIT:**
Better than creating a new sbt project with unit test is to **reuse the logic for finding projects that is in [build_tools/engine/context.rs](https://github.com/enso-org/enso/blob/9ac30c28e10fb9af7a6a5ca6d2716d56438c727e/build_tools/build/src/engine/context.rs#L588-L604)** that I have added for StdLibs API check. You can reuse that logic and basically replace `--docs api` commandline options with `--static-type-check`. With that, you can revert all the changes in sbt.

JaroslavTulach commented on engine/runtime/src/main/scala/org/enso/interpreter/runtime/util/DiagnosticFormatter.scala:
```diff
@@ -25,17 +29,48 @@ class DiagnosticFormatter(
+
+  sealed private trait DiagnosticKind
+
+  private object DiagnosticKind {
```
There are massive changes in this file!

radeusgd replied:
True. I will try to separate the logic a bit more.

JaroslavTulach commented on test/Exploratory_Benchmarks/package.yaml:
Thank you for cleaning this debris.

JaroslavTulach commented on test/Base_Tests/src/Runtime/Stack_Size_Spec.enso:
```diff
@@ -59,7 +60,7 @@ enso_bin -> File =
    - nesting_level How many times should the vector be nested
 generate_code nesting_level:Integer -> Text =
-    bldr = Vector.Builder.new
+    bldr = Vector_Builder.new
```
Why do we need this change? `Vector.Builder` seems like a correct way to reference the `Builder` when `Vector` is imported (which it is via `all` from base).

radeusgd replied:
Not really. I have no idea why it works at runtime to be honest.

The imported `Vector` symbol is the `Vector` **type** - it has static methods like `Vector.new` etc.

But the `Builder` however is declared on the `Vector` **module**.

The `Vector` type is also defined on `Vector` module. So if `Vector.Builder` is accessible, probably so should be `Vector.Vector`. IMO this is some bug in the runtime resolution.

This experiment seems to confirm it - in REPL I try typing:
```
> Vector.Vector.Vector.Vector
>>> Vector
```

I guess I should file a ticket to resolve this inconsistency.

radeusgd replied:
Reported #12946

radeusgd commented:
> Better than creating a new sbt project with unit test is to **reuse the logic for finding projects that is in [build_tools/engine/context.rs](https://github.com/enso-org/enso/blob/9ac30c28e10fb9af7a6a5ca6d2716d56438c727e/build_tools/build/src/engine/context.rs#L588-L604)** that I have added for StdLibs API check. You can reuse that logic and basically replace `--docs api` commandline options with `--static-type-check`. With that, you can revert all the changes in sbt.

We are moving towards migrating from the Rust build script to Bazel. I first started writing logic for detecting projects in Rust, but realized that will soon be migrated to Bazel. Sbt is not going away anytime soon so it seems like the right place. Moreover it provides devs an easy to run command - `sbt lintEnso` to run a check before submitting a PR and waiting for CI to do it.

Akirathan reviewed (approved):
Unblocking. The sbt magic seems like the lesser evil after all.

radeusgd commented:
> I don't like duplicating logic for finding projects inside sbt

Come on, the logic for finding projects is just a few lines of code. And it is specifically finding things for the lint check - not sure if the other one you mention is doing the same thing. Moreover I don't see where the same logic can be found in `runtime-integration-tests`, can you help me locate it?

>  I don't like the magic with aggregating the projects into a single one

This will be handled in #12937. If you insist I can do it as part of this PR, but it will only further delay everything.

radeusgd commented:
> Moreover, the `includeGithubAnnotation` functionality could then be moved from `DiagnosticFormatter` (where it does not belong), into `runtime-lints/test` sources.

I will try to refactor it a bit more to make this cleaner.

However it still is easiest to have Github logic somewhere next to `DiagnosticFormatter`. That is because this is where we have all the information for generating the annotations.

Another alternative is reporting the diagnostics in some common format (JSON?) and then parsing them to generate the annotations. But I think this is just overkill.

JaroslavTulach commented on engine/runtime/src/main/scala/org/enso/interpreter/runtime/util/DiagnosticFormatter.scala:
```diff
@@ -200,31 +323,81 @@ class DiagnosticFormatter(
-  ): Boolean = {
-    loc.end() <= source.getLength
+  private def includeGithubAnnotation: Boolean =
+    sys.env.get("GITHUB_ACTIONS").contains("true")
```
- can we somehow separate this "GitHub Actions" code from regular Enso CLI code?
- @Akirathan suggested to put it into completely different place, but I am afraid of that being too complex
- however a simple change would be to introduce `GitHubDiagnosticsFormatter extends DiagnosticsFormatter`
- and put this "GitHub Actions" code there.
- such a change would be a mere refactoring
- and it would keep the `DiagnosticsFormatter` almost simple

JaroslavTulach replied:
Re. [Radek's reply](https://github.com/enso-org/enso/pull/12901#issuecomment-28270527230):

> reporting the diagnostics in some common format (JSON?) and then parsing them to generate the annotations. But I think this is just overkill.

Yes, that is an overkill. Simple refactoring of the code will do.

radeusgd replied:
Yes that is a great suggestion, after @Akirathan I also considered extracting the GH-specific parts into a separate class, so I will try to do something like that.

I agree that this change to this file grew out of proportion and it is violating SRP now.

radeusgd replied:
Refactored in 48327a6b4132fff7c130c410e648a31d80b13276 (with some small followup changes)

JaroslavTulach replied:
I was hoping the refectoring could leave most of the `DiagnosticsFormatter` untouched, but alas, the changes a still too huge for me to review.

JaroslavTulach reviewed (commented):
- it'd be nice to limit the amount of `sbt` magic
- **request**: make the name of CLI option more standard and less verbose
- **advice**: separate _"GitHub Actions"_ formatter code into independent source

radeusgd committed:
Merge branch 'develop' into wip/rw/type-analysis-on-ci

radeusgd committed:
extract common logic for needs internals access

radeusgd committed:
CR

radeusgd committed:
revert _ -> * change

radeusgd committed:
remove unused import

radeusgd committed:
fmt

radeusgd committed:
CR: rename CLI arg to -Werror

radeusgd committed:
fix missing arg in test

radeusgd committed:
add test: checking inside of functions with default args should be done

radeusgd committed:
add test for type check for default arg

radeusgd committed:
TODO return type validation

radeusgd committed:
introduce 'type coverage' tool

radeusgd committed:
introduce 'type coverage' tool

fix

radeusgd committed:
postpone return in case of default args to process the function body

radeusgd committed:
WIP checking default args

radeusgd committed:
fix test

radeusgd committed:
Revert "introduce 'type coverage' tool"

This reverts commit d81425809e07f0169478c0e5c9be07905cbf4327.

radeusgd committed:
refactor diagnostic formatter

radeusgd committed:
Revert "introduce 'type coverage' tool"

This reverts commit ea302aa2a03bb7335940662dc443f76ac7918758.

radeusgd committed:
fix imports, add test

radeusgd committed:
add a test for Werror

radeusgd committed:
Merge branch 'develop' into wip/rw/type-analysis-on-ci

radeusgd commented:
Yet another finding from the type checker:
```
X:\NBO\enso\built-distribution\enso-engine-0.0.0-dev-windows-amd64\enso-0.0.0-dev\lib\Standard\Database\0.0.0-dev\src\DB_Table.enso:1146:40: warning: Invoking a value that has a non-function type (type Unsupported_Database_Operation) will result in a Not_Invokable error in runtime.
 1146 |             _ : Vector -> Error.throw (Unsupported_Database_Operation "`Vector` for `set`")
      |                                        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

```diff
diff --git a/distribution/lib/Standard/Database/0.0.0-dev/src/DB_Table.enso b/distribution/lib/Standard/Database/0.0.0-dev/src/DB_Table.enso
--- a/distribution/lib/Standard/Database/0.0.0-dev/src/DB_Table.enso	(revision 88110a98cb127563aaaf3e93231a863333b7eeb9)
+++ b/distribution/lib/Standard/Database/0.0.0-dev/src/DB_Table.enso	(date 1745838032241)
@@ -1143,7 +1143,7 @@
                     Error.throw (Integrity_Error.Error "Column "+value.name)
             _ : Constant_Column -> self.make_constant_column value
             _ : Simple_Expression -> value.evaluate self (set_mode==Set_Mode.Update && as=="") on_problems
-            _ : Vector -> Error.throw (Unsupported_Database_Operation "`Vector` for `set`")
+            _ : Vector -> Error.throw (Unsupported_Database_Operation.Error "`Vector` for `set`")
             _ : Array -> Error.throw (Unsupported_Database_Operation.Error "`Array` for `set`")
             _ : Range -> Error.throw (Unsupported_Database_Operation.Error "`Range` for `set`")
             _ : Date_Range -> Error.throw (Unsupported_Database_Operation.Error "`Date_Range` for `set`")

radeusgd committed:
fix discovered error

radeusgd committed:
more precise error location

radeusgd committed:
update test

radeusgd committed:
fmt

radeusgd committed:
Merge branch 'develop' into wip/rw/type-analysis-on-ci

JaroslavTulach commented on engine/runtime/src/main/scala/org/enso/interpreter/runtime/util/GitHubDiagnosticFormatter.scala:
```diff
@@ -0,0 +1,154 @@
+import scala.annotation.tailrec
+
+/** An extension of [[DiagnosticFormatter]] that additionally prints commands for a GitHub workflow that will add annotations for each warning/error. */
+class GitHubDiagnosticFormatter(
```
I know encapsulation isn't Scala's best attitude, but:
```suggestion
private[util] class GitHubDiagnosticFormatter(
```
would make me feel better. Isn't `DiagnosticsFormatter.make` (or rather `create`?) the right factory method to create an instance of formatter?

radeusgd replied:
That seems fair.

JaroslavTulach commented on engine/runtime/src/main/scala/org/enso/interpreter/runtime/util/DiagnosticFormatter.scala:
```diff
@@ -25,17 +26,46 @@ class DiagnosticFormatter(
+    case DiagnosticKind.Warning => fansi.Color.Yellow ++ fansi.Bold.On
+  }
+
+  private val subject: String = diagnosticKind match {
```
`val` create a field in the instance, doesn't it? Why do we need field `subject` when there already is a field `diagnosticKind` which is also redundant as the kind is derived based on type `diagnostics`.

These fields should be turned into methods unless there is some reason why it is better to trade some negliable CPU gain for memory.

JaroslavTulach commented on engine/runtime/src/main/scala/org/enso/interpreter/runtime/util/DiagnosticFormatter.scala:
```diff
@@ -25,17 +26,46 @@ class DiagnosticFormatter(
+
+  sealed protected trait DiagnosticKind
+
+  protected object DiagnosticKind {
```
Wouldn't it be better to switch to Java and just do:
```java
enum DiagnosticKind {
  ERROR, WARNING;
}
```
?

Just a note of despaired reader of Scala code....

radeusgd replied:
Yeah, I think you are right...

JaroslavTulach commented on engine/runtime/src/main/scala/org/enso/interpreter/runtime/util/DiagnosticFormatter.scala:
```diff
@@ -60,88 +90,142 @@ class DiagnosticFormatter(
   }
 
-  private def textAndLocation(): (String, SourceSection) = {
+  sealed protected trait Location {
```
`Location`? We already have a `Location` class elsewhere. Please choose a different name.

JaroslavTulach commented on engine/runner/src/main/java/org/enso/runner/Main.java:
```diff
@@ -92,6 +92,7 @@ public class Main {
   private static final String NO_READ_IR_CACHES_OPTION = "no-read-ir-caches";
   private static final String DISABLE_PRIVATE_CHECK_OPTION = "disable-private-check";
   private static final String ENABLE_STATIC_ANALYSIS_OPTION = "enable-static-analysis";
+  private static final String TREAT_WARNINGS_AS_ERRORS_OPTION = "Werror";
```
The option has been renamed. Thank you.

JaroslavTulach reviewed (approved):
- thanks for the `-Werror` option
- the `DiagnosticsFormatter` contains code changes I am more able to review now

radeusgd committed:
CR

radeusgd committed:
CR

radeusgd committed:
clean

radeusgd committed:
fix tests: logic to ignore GH diagnostics in some output

radeusgd committed:
Merge branch 'develop' into wip/rw/type-analysis-on-ci

radeusgd committed:
update bench processor to have matching namespace

radeusgd committed:
remove special cases in tests

radeusgd committed:
use separate env var for GH annotations - only add them in Lint check

radeusgd committed:
Merge branch 'develop' into wip/rw/type-analysis-on-ci

radeusgd commented:
I was getting many test failures on CI as our error messages started including the `::error` annotations for compile errors and the tests were not expecting that.

I started amending the tests to filter out these lines, but after enough trials I decided this is not the right approach - any new tests we will add would also need to be careful about this and it will be annoying. The worst part is - the tests will work locally and only fail with the `GITHUB_ACTIONS` env var set. This would be annoying to debug.

radeusgd commented:
So in d58c69b984e7c2683d7c4843171544cf4eec9c6c I changed it and now the GitHub annotation commands are only printed if a _special_ env var `ENSO_LINT_ENABLE_GITHUB_ANNOTATIONS` is set.

This env var is currently only set in the 'Enso Code Lint' workflow. So the regular engine/JVM tests will not be affected by changed error messages, making it much more predictable.

radeusgd committed:
test warnings

radeusgd committed:
rewrite locations

radeusgd committed:
Revert "test warnings"

This reverts commit f9881e9b586930a185aff74646a473ded71578ed.

radeusgd committed:
Merge branch 'develop' into wip/rw/type-analysis-on-ci

radeusgd committed:
fix missing new arg

JaroslavTulach commented on project/DistributionPackage.scala:
```diff
@@ -377,6 +375,9 @@ object DistributionPackage {
       all.add("--disable-private-check")
     }
     pb.command(all)
+    workingDirectory
```
CCing @AdRiley:
- I believe this change of workingDirectory breaks a lot of usual commands
- for example, as reported by Adam:
```
sbt:enso> runEngineDistribution --docs api --in-project distribution/lib/Standard/Tableau/0.0.0-dev
```
- was broken, in my opinion
- I am changing it in https://github.com/enso-org/enso/pull/12972/files#r2072659876

JaroslavTulach commented:
- FYI: #13067 claims that the Lint check is running in more jobs than it should

radeusgd commented:
> * FYI: [Linting passes should run only once in the CI #13067](https://github.com/enso-org/enso/issues/13067) claims that the Lint check is running in more jobs than it should

Explained there - I think this is a misunderstanding because 'lint' means multiple things. "Unused variable" check is a lint and it runs with _every_ compilation (and it's been like that forever and I think that's OK), and type inference is a lint that requires `--enable-static-analysis` to run. Unless some changes have been done in my absence, I highly doubt that this is running in normal workflows, it should only be run in a workflow that explicitly enables it.
